File: mobile/src/services/autocompleteHistoryService.ts
/**
 * Service pour gérer l'historique des caractéristiques autocomplete
 * Permet de récupérer les suggestions intelligentes depuis le backend
 */

import { apiCall, apiGet } from './api';

export interface AutocompleteSuggestion {
    valeur: string;
    usage_count: number;
}

export interface AutocompleteCharacteristic {
    id: number;
    identifiant_base: string;
    sous_caracteristique: string;
    valeur: string;
    origine_champs: string;
    user_id?: number;
    service_id?: number;
    usage_count: number;
    created_at: string;
    updated_at: string;
}

class AutocompleteHistoryService {
    /**
     * Récupérer les suggestions autocomplete pour une sous-caractéristique
     * @param identifiant_base Ex: "caracteristiques_vehicule"
     * @param sous_caracteristique Ex: "marque", "modele", "annee"
     * @param prefix Préfixe de recherche (optionnel)
     * @param limit Nombre maximum de suggestions (défaut: 10)
     */
    async getSuggestions(
        identifiant_base: string,
        sous_caracteristique: string,
        prefix?: string,
        limit: number = 10
    ): Promise<AutocompleteSuggestion[]> {
        try {
            const params = new URLSearchParams({
                identifiant_base,
                sous_caracteristique,
                limit: limit.toString(),
            });

            if (prefix) {
                params.append('prefix', prefix);
            }

            const response = await apiGet<{
                success: boolean;
                data: AutocompleteSuggestion[];
                count: number;
            }>(`/api/autocomplete/suggestions?${params.toString()}`);

            if (response.success && Array.isArray(response.data)) {
                return response.data;
            }

            return [];
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur récupération suggestions:', error);
            return [];
        }
    }

    /**
     * Récupérer toutes les sous-caractéristiques disponibles pour un identifiant_base
     * @param identifiant_base Ex: "caracteristiques_vehicule"
     */
    async getSubCharacteristics(identifiant_base: string): Promise<string[]> {
        try {
            const response = await apiGet<{
                success: boolean;
                data: string[];
                count: number;
            }>(`/api/autocomplete/sub-characteristics/${encodeURIComponent(identifiant_base)}`);

            if (response.success && Array.isArray(response.data)) {
                return response.data;
            }

            return [];
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur récupération sous-caractéristiques:', error);
            return [];
        }
    }

    /**
     * Récupérer toutes les valeurs pour une combinaison identifiant_base + sous_caracteristique
     * @param identifiant_base Ex: "caracteristiques_vehicule"
     * @param sous_caracteristique Ex: "marque"
     */
    async getAllValues(
        identifiant_base: string,
        sous_caracteristique: string
    ): Promise<string[]> {
        try {
            const response = await apiGet<{
                success: boolean;
                data: string[];
                count: number;
            }>(
                `/api/autocomplete/values/${encodeURIComponent(identifiant_base)}/${encodeURIComponent(sous_caracteristique)}`
            );

            if (response.success && Array.isArray(response.data)) {
                return response.data;
            }

            return [];
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur récupération valeurs:', error);
            return [];
        }
    }

    /**
     * Insérer ou mettre à jour une caractéristique autocomplete
     * @param identifiant_base Ex: "caracteristiques_vehicule"
     * @param sous_caracteristique Ex: "marque"
     * @param valeur Ex: "Toyota"
     * @param origine_champs "ia" ou "utilisateur"
     * @param user_id ID de l'utilisateur (optionnel)
     * @param service_id ID du service (optionnel)
     */
    async upsertCharacteristic(
        identifiant_base: string,
        sous_caracteristique: string,
        valeur: string,
        origine_champs: 'ia' | 'utilisateur' = 'utilisateur',
        user_id?: number,
        service_id?: number
    ): Promise<number | null> {
        try {
            const response = await apiCall<{
                success: boolean;
                id: number;
                message: string;
            }>('/api/autocomplete/upsert', {
                method: 'POST',
                body: JSON.stringify({
                    identifiant_base,
                    sous_caracteristique,
                    valeur,
                    origine_champs,
                    user_id,
                    service_id,
                }),
            });

            if (response.success && response.data?.id) {
                return response.data.id;
            }

            return null;
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur upsert caractéristique:', error);
            return null;
        }
    }

    /**
     * Historiser un champ autocomplete complet
     * Découpe les valeurs concaténées et historise chaque sous-caractéristique
     * @param identifiant_base Ex: "caracteristiques_vehicule"
     * @param valeurs Tableau de valeurs concaténées Ex: ["Toyota,RAV4,2018,4x4"]
     * @param separateur Séparateur utilisé (défaut: ",")
     * @param sous_caracteristiques Objet avec les noms des sous-caractéristiques
     * @param origine_champs "ia" ou "utilisateur"
     * @param user_id ID de l'utilisateur (optionnel)
     * @param service_id ID du service (optionnel)
     */
    async historizeField(
        identifiant_base: string,
        valeurs: string[],
        separateur: string,
        sous_caracteristiques: Record<string, string[]>,
        origine_champs: 'ia' | 'utilisateur' = 'utilisateur',
        user_id?: number,
        service_id?: number
    ): Promise<number[]> {
        try {
            const response = await apiCall<{
                success: boolean;
                ids: number[];
                count: number;
                message: string;
            }>('/api/autocomplete/historize', {
                method: 'POST',
                body: JSON.stringify({
                    identifiant_base,
                    valeurs,
                    separateur,
                    sous_caracteristiques,
                    origine_champs,
                    user_id,
                    service_id,
                }),
            });

            if (response.success && Array.isArray(response.data?.ids)) {
                return response.data.ids;
            }

            return [];
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur historisation:', error);
            return [];
        }
    }

    /**
     * Historiser automatiquement un champ autocomplete depuis les données d'un produit
     * Cette fonction est appelée lors de la création d'un produit pour enrichir l'historique
     * @param fieldData Données du champ autocomplete depuis le formulaire
     * @param user_id ID de l'utilisateur (optionnel)
     * @param service_id ID du service (optionnel)
     */
    async historizeFromFieldData(
        fieldData: {
            identifiant_base?: string;
            valeur?: string[] | string;
            separateur?: string;
            sous_caracteristiques?: Record<string, string[]>;
            origine_champs?: string;
        },
        user_id?: number,
        service_id?: number
    ): Promise<void> {
        try {
            // Vérifier que c'est bien un champ autocomplete
            if (!fieldData.identifiant_base || !fieldData.sous_caracteristiques) {
                return;
            }

            // Normaliser les valeurs
            const valeurs = Array.isArray(fieldData.valeur)
                ? fieldData.valeur
                : fieldData.valeur
                    ? [fieldData.valeur]
                    : [];

            if (valeurs.length === 0) {
                return;
            }

            const separateur = fieldData.separateur || ',';
            const origine_champs = (fieldData.origine_champs as 'ia' | 'utilisateur') || 'utilisateur';

            // Historiser le champ complet
            await this.historizeField(
                fieldData.identifiant_base,
                valeurs,
                separateur,
                fieldData.sous_caracteristiques,
                origine_champs,
                user_id,
                service_id
            );

            console.log(
                `[AutocompleteHistoryService] ✅ Champ historisé: ${fieldData.identifiant_base} (${valeurs.length} valeurs)`
            );
        } catch (error) {
            console.error('[AutocompleteHistoryService] Erreur historisation depuis fieldData:', error);
        }
    }
}

export const autocompleteHistoryService = new AutocompleteHistoryService();

File: backend/src/services/autocomplete_history_service.rs
// Service pour gérer l'historique des caractéristiques autocomplete
use crate::core::types::AppError;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use sqlx::Row;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AutocompleteCharacteristic {
    pub id: i32,
    pub identifiant_base: String,
    pub sous_caracteristique: String,
    pub valeur: String,
    pub origine_champs: String,
    pub user_id: Option<i32>,
    pub service_id: Option<i32>,
    pub usage_count: i32,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AutocompleteSuggestion {
    pub valeur: String,
    pub usage_count: i32,
}

/// Insérer ou mettre à jour une caractéristique autocomplete
pub async fn upsert_autocomplete_characteristic(
    pool: &PgPool,
    identifiant_base: &str,
    sous_caracteristique: &str,
    valeur: &str,
    origine_champs: &str,
    user_id: Option<i32>,
    service_id: Option<i32>,
) -> Result<i32, AppError> {
    log::info!(
        "[AutocompleteHistoryService] Upsert caractéristique: {} > {} = {}",
        identifiant_base,
        sous_caracteristique,
        valeur
    );

    // Utiliser la fonction SQL upsert_autocomplete_characteristic
    let row = sqlx::query(
        r#"
        SELECT upsert_autocomplete_characteristic(
            $1, $2, $3, $4, $5, $6
        ) as id
        "#,
    )
    .bind(identifiant_base)
    .bind(sous_caracteristique)
    .bind(valeur)
    .bind(origine_champs)
    .bind(user_id)
    .bind(service_id)
    .fetch_one(pool)
    .await
    .map_err(|e| AppError::Internal(format!("Erreur upsert caractéristique: {}", e)))?;

    let id: i32 = row.get::<i32, _>("id");
    log::info!(
        "[AutocompleteHistoryService] ✅ Caractéristique sauvegardée: {}",
        id
    );

    Ok(id)
}

/// Historiser plusieurs caractéristiques autocomplete depuis un champ autocomplete
pub async fn historize_autocomplete_field(
    pool: &PgPool,
    identifiant_base: &str,
    valeurs: &[String],
    separateur: &str,
    sous_caracteristiques: &serde_json::Value,
    origine_champs: &str,
    user_id: Option<i32>,
    service_id: Option<i32>,
) -> Result<Vec<i32>, AppError> {
    log::info!(
        "[AutocompleteHistoryService] Historisation champ autocomplete: {} ({} valeurs)",
        identifiant_base,
        valeurs.len()
    );

    let mut ids = Vec::new();

    // Extraire les noms des sous-caractéristiques depuis le JSON
    let sous_caracs_obj = if let Some(obj) = sous_caracteristiques.as_object() {
        obj
    } else {
        log::warn!("[AutocompleteHistoryService] sous_caracteristiques n'est pas un objet");
        return Ok(ids);
    };

    // Pour chaque valeur concaténée, la découper et historiser chaque sous-caractéristique
    for valeur_concat in valeurs {
        let parts: Vec<&str> = valeur_concat.split(separateur).map(|s| s.trim()).collect();

        // Historiser chaque sous-caractéristique
        for (idx, sous_carac_name) in sous_caracs_obj.keys().enumerate() {
            if let Some(valeur_part) = parts.get(idx) {
                if !valeur_part.is_empty() {
                    match upsert_autocomplete_characteristic(
                        pool,
                        identifiant_base,
                        sous_carac_name,
                        valeur_part,
                        origine_champs,
                        user_id,
                        service_id,
                    )
                    .await
                    {
                        Ok(id) => ids.push(id),
                        Err(e) => {
                            log::warn!(
                                "[AutocompleteHistoryService] Erreur historisation {}: {}",
                                sous_carac_name,
                                e
                            );
                        }
                    }
                }
            }
        }
    }

    log::info!(
        "[AutocompleteHistoryService] ✅ {} caractéristiques historisées",
        ids.len()
    );
    Ok(ids)
}

/// Récupérer les suggestions pour une sous-caractéristique donnée
pub async fn get_autocomplete_suggestions(
    pool: &PgPool,
    identifiant_base: &str,
    sous_caracteristique: &str,
    prefix: Option<&str>,
    limit: i64,
) -> Result<Vec<AutocompleteSuggestion>, AppError> {
    let rows = if let Some(prefix) = prefix {
        sqlx::query(
            r#"
            SELECT valeur, usage_count
            FROM autocomplete_characteristics
            WHERE identifiant_base = $1
            AND sous_caracteristique = $2
            AND LOWER(valeur) LIKE LOWER($3 || '%')
            ORDER BY usage_count DESC, valeur ASC
            LIMIT $4
            "#,
        )
        .bind(identifiant_base)
        .bind(sous_caracteristique)
        .bind(prefix)
        .bind(limit)
        .fetch_all(pool)
        .await
    } else {
        sqlx::query(
            r#"
            SELECT valeur, usage_count
            FROM autocomplete_characteristics
            WHERE identifiant_base = $1
            AND sous_caracteristique = $2
            ORDER BY usage_count DESC, valeur ASC
            LIMIT $3
            "#,
        )
        .bind(identifiant_base)
        .bind(sous_caracteristique)
        .bind(limit)
        .fetch_all(pool)
        .await
    }
    .map_err(|e| AppError::Internal(format!("Erreur récupération suggestions: {}", e)))?;

    let suggestions = rows
        .iter()
        .map(|row| AutocompleteSuggestion {
            valeur: row.get::<String, _>("valeur"),
            usage_count: row.get::<i32, _>("usage_count"),
        })
        .collect();

    Ok(suggestions)
}

/// Récupérer toutes les sous-caractéristiques disponibles pour un identifiant_base
pub async fn get_sub_characteristics(
    pool: &PgPool,
    identifiant_base: &str,
) -> Result<Vec<String>, AppError> {
    let rows = sqlx::query(
        r#"
        SELECT DISTINCT sous_caracteristique
        FROM autocomplete_characteristics
        WHERE identifiant_base = $1
        ORDER BY sous_caracteristique ASC
        "#,
    )
    .bind(identifiant_base)
    .fetch_all(pool)
    .await
    .map_err(|e| AppError::Internal(format!("Erreur récupération sous-caractéristiques: {}", e)))?;

    let sub_chars: Vec<String> =
        rows.iter().map(|row| row.get::<String, _>("sous_caracteristique")).collect();
    Ok(sub_chars)
}

/// Récupérer toutes les valeurs pour une combinaison identifiant_base + sous_caracteristique
pub async fn get_all_values(
    pool: &PgPool,
    identifiant_base: &str,
    sous_caracteristique: &str,
) -> Result<Vec<String>, AppError> {
    let rows = sqlx::query(
        r#"
        SELECT DISTINCT valeur
        FROM autocomplete_characteristics
        WHERE identifiant_base = $1
        AND sous_caracteristique = $2
        ORDER BY usage_count DESC, valeur ASC
        "#,
    )
    .bind(identifiant_base)
    .bind(sous_caracteristique)
    .fetch_all(pool)
    .await
    .map_err(|e| AppError::Internal(format!("Erreur récupération valeurs: {}", e)))?;

    let values: Vec<String> = rows.iter().map(|row| row.get::<String, _>("valeur")).collect();
    Ok(values)
}

/// Supprimer une caractéristique (pour nettoyage)
pub async fn delete_autocomplete_characteristic(pool: &PgPool, id: i32) -> Result<bool, AppError> {
    let result = sqlx::query("DELETE FROM autocomplete_characteristics WHERE id = $1")
        .bind(id)
        .execute(pool)
        .await
        .map_err(|e| AppError::Internal(format!("Erreur suppression caractéristique: {}", e)))?;

    Ok(result.rows_affected() > 0)
}
